File: Project_1/src/main/java/it/uniroma2/sabd/engineering/SparkSingleton.java
package it.uniroma2.sabd.engineering;

import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;
import org.apache.spark.sql.SparkSession;

import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

public class SparkSingleton {

    private final SparkContext sparkContext ;
    private final SparkSession sparkSession ;
    private static SparkSingleton instance;

    private SparkSingleton()  {

        try (InputStream input = getClass().getClassLoader().getResourceAsStream("spark.properties")) {

            Properties sparkProperties = new Properties() ;
            sparkProperties.load(input);

            SparkConf conf = new SparkConf()
                    .setMaster(sparkProperties.getProperty("spark.master"))
                    .setAppName(sparkProperties.getProperty("spark.appName")) ;

            sparkContext = new SparkContext(conf);
            sparkSession = SparkSession.builder()
                    .config(conf)
                    .getOrCreate() ;
        }

        catch (IOException e) {
            throw new RuntimeException(e);
        }

    }

    public static SparkSingleton getInstance() {
        if (instance == null) {
            instance = new SparkSingleton() ;
        }
        return instance;
    }

    public SparkContext getSparkContext() {
        return this.sparkContext ;
    }

    public SparkSession getSparkSession() {
        return this.sparkSession ;
    }
}

File: src/engineering/SparkSingleton.py
import jproperties as jprop
from pyspark import *
from pyspark.sql import *


__sparkContext = None
__sparkSession = None


def getSparkContext() -> SparkContext :
    global __sparkContext
    if (__sparkContext == None) :
        configs = jprop.Properties()

        with open("./properties/spark.properties", 'rb') as config_file:
            configs.load(config_file)

            masterName = configs.get("spark.master").data
            appName = configs.get("spark.appName").data
            masterPort = configs.get("spark.port").data

            sparkMasterUrl = "spark://" + masterName + ":" + masterPort
            
            __sparkContext = SparkContext(
                master = sparkMasterUrl , 
                appName = appName, 
            )
    return __sparkContext


def getSparkSession() -> SparkSession :
    global __sparkSession
    if (__sparkSession == None) :
        sparkContext = getSparkContext()
        __sparkSession = SparkSession(sparkContext = sparkContext)
    return __sparkSession


def resetConnection() :
    global __sparkContext
    global __sparkSession

    if (__sparkSession != None) :
        __sparkSession.stop()
        __sparkSession = None

    if (__sparkContext != None) :
        __sparkContext.stop()
        __sparkContext = None